src/pyramid/core.cljc

```diff
@@ -53,6 +53,9 @@
 (defn- replace-all-nested-entities
   [identify v]
   (cond
+    ;; Don't normalize records
+    (record? v) v
+
```

Comment:
I'm not so sure this is right yet. I'm not sure it's wrong yet either.

The root cause of #17 is that in `main` this takes the `(map? v)` path which calls `(empty v)`. In Clojure this is an error, it seems that in ClojureScript this returns `nil`.

Another potential fix would be to appropriately consume and reconstruct the record here. I've been reading the source for `clojure.walk` a lot recently and the way they do it is by using the record `v` as the initial arg, e.g.

```suggestion
    ;; Normalize records while preserving type
    ;; TODO could we move this into the `map?` branch and just change what we pass in to `into`
    (record? v) (into v (map (juxt first (comp #(replace-all-nested-entities identify %) second))) v)

```
I haven't tested this yet but something akin to that should preserve the type of `v` while also allowing us to descend into it and continue normalizing. In the case of the lambdaisland/uri data, I expect it wouldn't change anything about it unless you configured your `identify` function to do something special with the shape of URI records. What do you think?

Reply:
Interesting! But isn't the semantics of normalizing something from inside a record rather dubious?

Ie., if Pyramid finds something inside the record that looks like an entity, is it correct to modify the record and leave a reference in its place? I'd guess that it would very, very rarely be correct. And if it's not correct, what's the point of descending into them? Unlike Walk, Pyramid has pretty clear semantics, and doesn't really need to do general purpose processing over arbitrary structures.

Reply:
I agree it's a bit nebulous at this point. I am not a heavy user of records, and so it's hard for me to anticipate how people might want to use them with pyramid.

Since I have to choose, I think I am leaning towards treating records like maps and normalizing them and their contents, since the point of records is that you have something that acts like clojure data but can be extended with additional protocols.

I think it's unlikely most records would contain information that should be normalized, but if they do, it would be unfortunate as a user of the library to be unable to pierce that veil.

Reply:
I think the opposite is true as well: if the expectation is that the record subscribes to a certain shape, it would be unfortunate to have that shape changed in an unexpected manner. 

Given that records behave like a type (but without having to create a special mini-DSL around it), enumerate their fields, and can have those fields type hinted, if they do take a map as one of the fields, I think it's unlikely that any protocol it would also expect to find an ident in that same field. 

If you do have a record that incidentally have something that could be normalized in one of its fields, and you want to extract it, intact, after it's been through Pyramid, how would you do it?

I think an expectation of records is that you can create something that is named, doesn't require a special DSL, but also has integrity. I think pushing the normalization straight through them will break the last part of that promise.

Reply:
`pull` could certainly recursively reconstruct descendants of records, although the type of it would depend on the user asking for all of the requisite fields of the record in the query else it would have to return a map, just like dissocing a field from a record returns a map.

Perhaps there's a third way - could we create an API that would allow a user to opt in or out of the normalization process per type? I could see this done with a protocol that the record can implement if they choose to opt in, e.g.

```clojure
(ns pyramid.core)

(defprotocol EntityRecord)

,,,,


(ns foo.bar
  (:require
   [pyramid.core :as p]))

;; opt in to normalization
(defrecord User [id name friends]
  p/EntityRecord
  ,,,)

;; this would not be normalized, because it does not implement p/EntityRecord
(defrecord SystemOperation [id status])
```

Reply:
Opting in to normalization is an intriguing idea. To me, that would be the unsurprising option, that still allows for the functionality.